File: ffi/examples/get_started/src/run_example.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include <tvm/ffi/container/ndarray.h>
#include <tvm/ffi/extra/module.h>

// This file shows how to load the same compiled module and interact with it in C++
namespace ffi = tvm::ffi;

struct CPUNDAlloc {
  void AllocData(DLTensor* tensor) { tensor->data = malloc(ffi::GetDataSize(*tensor)); }
  void FreeData(DLTensor* tensor) { free(tensor->data); }
};

inline ffi::NDArray Empty(ffi::Shape shape, DLDataType dtype, DLDevice device) {
  return ffi::NDArray::FromNDAlloc(CPUNDAlloc(), shape, dtype, device);
}

int main() {
  // load the module
  ffi::Module mod = ffi::Module::LoadFromFile("build/add_one_cpu.so");

  // create an NDArray, alternatively, one can directly pass in a DLTensor*
  ffi::NDArray x = Empty({5}, DLDataType({kDLFloat, 32, 1}), DLDevice({kDLCPU, 0}));
  for (int i = 0; i < 5; ++i) {
    reinterpret_cast<float*>(x->data)[i] = static_cast<float>(i);
  }

  ffi::Function add_one_cpu = mod->GetFunction("add_one_cpu").value();
  add_one_cpu(x, x);

  std::cout << "x after add_one_cpu(x, x)" << std::endl;
  for (int i = 0; i < 5; ++i) {
    std::cout << reinterpret_cast<float*>(x->data)[i] << " ";
  }
  std::cout << std::endl;
  return 0;
}

File: ffi/examples/packaging/run_example.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations.
# Base logic to load library for extension package
import torch
import sys
import my_ffi_extension


def run_add_one():
    x = torch.tensor([1, 2, 3, 4, 5], dtype=torch.float32)
    y = torch.empty_like(x)
    my_ffi_extension.add_one(x, y)
    print(y)


def run_raise_error():
    my_ffi_extension.raise_error("This is an error")


if __name__ == "__main__":
    if len(sys.argv) > 1:
        if sys.argv[1] == "add_one":
            run_add_one()
        elif sys.argv[1] == "raise_error":
            run_raise_error()
    else:
        print("Usage: python run_example.py <add_one|raise_error>")
